### PR TITLE
Add new AWS resource type and filter for ASG scaling-policies

### PR DESCRIPTION
**Pull request by al3pht**
# What's In This Pull Request
This pull request includes the addition of an AWS resource type ```scaling-policies``` which will return metadata for ASG scaling policies via the ```describe_scaling_policies``` boto3 API call. No additional filters or actions were added to this resource type outside of those that come out-of-the-box with custodian.  
It also includes the addition of the ```filter: scaling-policies``` for ```resource: asg``` which returns ASG metadata filtered by key and value for their scaling-policies e.g.:  
> ```
> filters:
>   type: scaling-policies
>   key: "KeyFromDescribeScalingPolicies"
>   value: "ValueFromDescribeScalngPolicies"  
We have also included a developer guide for contributing new resources to custodian similar to those that already exist for Azure and GCP
# Type of Change
- [x] Enhancement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
# Checklist
- [x] Added scaling-policies as a resource type
- [x] Added scaling-policies as a filter for asg
- [x] Added and passed tests in ```test_asg.py``` for both of the above
- [x] Built documentation ensuring changes to ```autoscaling``` were reflected
- [x] Added developer guide documentation for contributing to AWS

### CODE REVIEW

**Comment by kapilt** [comment]
Thanks for the pull request and additional details!  If you don't mind could you and the other committers here sign the CLA per the developer docs? Direct link here: https://docs.google.com/forms/d/e/1FAIpQLSfwtl1s6KmpLhCY6CjiY8nFZshDwf_wrmNYx1ahpsNFXXmHKw/viewform

**Comment by al3pht** [comment]
Hello Kapil,
It was a pleasure to contribute. Please let us know what changes we need to
make (if any). We are looking forward to contributing more. I have
submitted the CLA and I have notified the other committers so they can do
the same.
Regards

File: c7n/resources/asg.py
Comment (kapilt) [inline c7n/resources/asg.py:1756]: its unclear what the reuse potential for this is, we optimize for removability, not unused abstractions. i would suggest just merging this class into the filter below, we can refactor it back out if there are additional usages.

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:3]: :heart_eyes:  Thanks!

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:6]: we don't support python2, we also support MacOS

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:7]: 3.7 at this point, 3.6 is EOL in a year.

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:18]: ? if you have a developer install per make install, and the virtualenv activate, executing the source directly versus using the cli entrypoint is superflous, yes, its possible, but people should just use the cli `custodian` entrypoint since its the same workflow and command across usages instead of introducing additional one off extraneous mechanisms into the docs.

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:48]: speling typo

File: docs/source/aws/contribute.rst
Comment (kapilt) [inline docs/source/aws/contribute.rst:38]: so the actual documentation reference and which this content should be added to is the TypeInfo for query.py,

ideal is if we make it a class doc string there and then pull the doc here via sphinx.

**Review by kapilt — COMMENTED** [review]
looks good, a few minors per comments.

**Comment by al3pht** [comment]
Hello Kapil,
We'll send a new PR with modifications based on your feedback. Thank you
for the quick response.
Regards

**Comment by kapilt** [comment]
> Hello Kapil, We'll send a new PR with modifications based on your feedback. Thank you for the quick response. Regards
> […](#)

thanks, just to clarify, updates to this pr in place via push to branch is preferred over creating a new pr on GitHub.

**Comment by al3pht** [comment]
Understood, thanks

File: docs/source/aws/contribute.rst
Comment (liatgold) [reply docs/source/aws/contribute.rst:38]: Hi Kapil, thanks for all of your comments! 
I have not been able to find any references to query.py anywhere else in the documentation. Could you provide more information on how to make query.py a class docstring so it can be referenced elsewhere using sphinx?

Alternatively, in our updated PR we included the documentation you mentioned from query.py as a code block in our contribute.rst. Let us know if it this ok or if you'd rather we follow your initial suggestion.

Thanks,
Liat

Comment (kapilt) [reply docs/source/aws/contribute.rst:38]: i guess i was looking to add the docs that are in this new developer doc, back to query.py ie wrt to the structure of say enum_spec and detail_spec.. for those changes to be useful in this guide it would need a refactor of the docs in query.py to use a class doc string, and then include them back to this document via sphinx directive.

Comment (kapilt) [reply docs/source/aws/contribute.rst:38]: rethinking this, i think we should probably separate out the dev docs into a separate pr, as the rest of this looks close to ready to go, but the docs are going to need a bit more cleanup and are unrelated to the rest of this feature pr.

Comment (liatgold) [reply docs/source/aws/contribute.rst:38]: Hi Kapil - this PR has been updated (AWS dev docs removed).
Once the dev docs are ready, we'll submit a new PR.

**Comment by al3pht** [comment]
@kapilt we have removed the documentation code (we will submit in a different PR). Please let us know what else shall be modified, thx

**Comment by al3pht** [comment]
Hi @kapilt is there anything pending to review in this PR? All previous observations/comments have been addressed.

**Comment by kapilt** [comment]
> Hi @kapilt is there anything pending to review in this PR? All previous observations/comments have been addressed.

passing CI checks, tests look like their failing

File: c7n/resources/asg.py
Comment (kapilt) [inline c7n/resources/asg.py:1779]: auto scaling isn't supported by config so config_type should be removed here, this is currently triggering a test failure.

Comment (al3pht) [reply c7n/resources/asg.py]: Done

**Comment by kapilt** [comment]
it appears this pr is still failing ci, wrt to doc formatting, see makefile sphinx/doc targets re reproducing locally.

**Comment by liatgold** [comment]
Hi Kaplit - we fixed the issue and it looks like all the checks passed. Please let us know if there is anything else that needs to get done in order to approve the PR

**Comment by al3pht** [comment]
Hello @kapilt is there anything else pending in this PR?

**Comment by kapilt** [comment]
thanks for your patience, overall it looks good, there's some minors i was trying to walk through, but i can correct pre-merge.

**Comment by al3pht** [comment]
awesome!

**Comment by al3pht** [comment]
Hello @kapilt let us know if you'd like us to take a look at those "minors" so you don't have to correct at pre-merge. Thx

**Comment by al3pht** [comment]
hi @kapilt any updates? Once we have this PR merged we will be submitting other for the AWS contribution documentation.

File: c7n/resources/asg.py
Comment (kapilt) [inline c7n/resources/asg.py:1831]: initialize is returning self so this is creating a cycle memory reference, simply using the instance attribute dict of self.policies directly would be better.

**Comment by kapilt** [comment]
committers here pre-date easycla usage, and are recorded in previous cla forms.

**Comment by al3pht** [comment]
@kapilt do we have to enroll/use `easycla`? Do we need this for incoming PRs?
Please advise, thx

**Comment by kapilt** [comment]
@al3pht you don't need it for this pr, i'm going to white list as our previous cla was signed, new prs will need easycla usage, we're talking to them (linux foundation who runs the easy cla) about smoothing out some of the rough edges, their in progress on rolling out a major update this month which should address some of the pain points.

**Review by kapilt — APPROVED** [review]
lgtm, thanks

**Pull request merged**
